File: src/main/java/spacesimulation/Simulator.java
package spacesimulation;

import java.awt.*;
import java.awt.image.BufferStrategy;
import java.security.Key;
import java.util.ArrayList;

public class Simulator implements Runnable {
    private Thread thread;
    private boolean running;
    private final Display display;
    private BufferStrategy bs;
    private Graphics g;
    private final KeyManager keymanager = new KeyManager();
    private final ArrayList<Simulation> simulations;

    public Simulator() {
        display = new Display();
        display.getJFrame().addKeyListener(keymanager);
        simulations = new ArrayList<>();
    }

    private void tick(double dt) {
        for (Simulation simulation : simulations) simulation.parentTick(dt);
        keymanager.tick();
    }

    private void render() {
        bs = display.getCanvas().getBufferStrategy();
        if(bs == null) {
            display.getCanvas().createBufferStrategy(3);
            return;
        }
        g = bs.getDrawGraphics();
        g.clearRect(0, 0, display.getCanvas().getWidth(), display.getCanvas().getHeight());
        for (Simulation simulation : simulations) simulation.parentRender(g);
        bs.show();
        g.dispose();
    }

    public void addSimulation(Simulation simulation) {
        simulations.add(simulation);
        simulation.keyManager = keymanager;
    }

    @Override
    public void run() {
        init();
        long lastTime = System.currentTimeMillis();
        final double amountOfTicks = 160;
        double msPerTick = 1000.0 / amountOfTicks;
        double delta = 0;

        while (running) {
            long now = System.currentTimeMillis();
            delta += (now - lastTime) / msPerTick;
            if (delta >= 1) {
                tick((now - lastTime) / 1000.0);
                render();
                delta--;
                lastTime = now;
            }
        }
        stop();
    }

    private void init() {
        
    }

    public synchronized void start() {
        if(running) return;
        running = true;
        thread = new Thread(this);
        thread.start();
    }

    public synchronized void stop() {
        if(!running) return;
        running = false;
        try {
            thread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public int getCenterX() {
        return display.getCanvas().getWidth() / 2;
    }

    public int getCenterY() {
        return display.getCanvas().getHeight() / 2;
    }

    public int getHeight() {
        return display.getCanvas().getHeight();
    }

    public int getWidth() {
        return display.getCanvas().getWidth();
    }

    public KeyManager getKeymanager() {
        return keymanager;
    }
}

File: src/main/java/spacesimulation/Simulator.kt
package spacesimulation

import spacesimulation.physics.Seconds
import java.awt.Graphics
import java.awt.image.BufferStrategy

/**
 * Environment in which a simulation is running.
 */
class Simulator {
    val keyManager = KeyManager()
    private var threadTickingAndRendering: Thread? = null
    private var running = false
    private val display: Display = Display().apply{ jFrame.addKeyListener(keyManager) }
    private var bs: BufferStrategy? = null
    private var g: Graphics? = null
    private val simulations: ArrayList<Simulation> = ArrayList()

    /**
     * triggers [Simulation.render] on every simulation at a given frequency.
     * triggers [Simulation.tick] as often as possible
     */
    private val tickAndRender = Runnable {
        var lastTime = System.currentTimeMillis()
        val amountOfTicks = 160.0
        val msPerTick = 1000.0 / amountOfTicks
        var delta = 0.0
        while (running) {
            val now = System.currentTimeMillis()

            // always tick
            val dt: Seconds = (now - lastTime) / 1000.0
            tick(dt)
            delta += (now - lastTime) / msPerTick

            // render to reach fps goal
            if (delta >= 1) {
                render()
                delta--
                lastTime = now
            }
        }
        stop()
    }


    /**
     * Calls [Simulation.tick] on the simulations run in this simulator
     */
    private fun tick(dt: Double) {
        for (simulation in simulations) simulation.parentTick(dt)
        keyManager.tick()
    }

    /**
     * Calls [Simulation.render] with a new [Graphics] object
     */
    private fun render() {
        bs = display.canvas.bufferStrategy
        if (bs == null) {
            display.canvas.createBufferStrategy(3)
            return
        }
        g = bs!!.drawGraphics
        g!!.clearRect(0, 0, display.canvas.width, display.canvas.height)
        for (simulation in simulations) simulation.parentRender(g!!)
        bs!!.show()
        g!!.dispose()
    }

    /**
     * Registers a new [Simulation] in the Simulator
     * @param simulation the new Simulation
     */
    fun addSimulation(simulation: Simulation) {
        simulations.add(simulation)
        simulation.keyManager = keyManager
    }

    @Synchronized
    fun start() {
        if (running) return
        running = true
        threadTickingAndRendering = Thread(tickAndRender).apply {start()}
    }

    @Synchronized
    fun stop() {
        if (!running) return
        running = false
        try {
            threadTickingAndRendering?.join()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    val height: Int
        get() = display.canvas.height
    val width: Int
        get() = display.canvas.width
}